### DISCUSSION TITLE
[Feature Request] Preserve hwdef in DCP file

### DISCUSSION THREAD

**Discussion by nqdtan** [Xilinx/RapidWright#597] [General]
RapidWright currently does not preserve a `*.hwdef` file (if it exists) in a Design checkpoint file. What happened was that I imported a DCP file produced by RapidWright to Vivado in order to generate a bitstream with the command `write_device_image`. I then met the following error message in Vivado

> ERROR: [DRC CIPS-2] Versal CIPS exists check - wdi: Versal designs must contain a CIPS IP in the netlist hierarchy to function properly. Please create an instance of the CIPS IP and configure it. Without a CIPS IP in the design, Vivado will not generate a CDO for the PMC, an elf for the PLM.

By some investigations, I found that this error is due to the missing of a `hwdef` file in the DCP. I tried adding one (from the original design checkpoint) manually by compressing a new DCP with `zip` (plus inserting a **HWDEF** entry to the `dcp.xml` file) and was able to generate bitstream. Attached is a DCP sample generated by Vitis 2021.2 that has a `hwdef` file.

I think it would be convenient if future RapidWright version preserves `hwdef` file when writing out DCP.
[DCP Sample with HWDEF](https://drive.google.com/file/d/1zqnw30PaUYN3xJLEvVwu4j5iB8H3GCVo/view?usp=sharing)

**Comment by clavin-xlnx** [comment]
Thanks @nqdtan - I have a fix in the works and it will be in the next release.  I will reference this post when it is available.

**Reply by clavin-xlnx** [reply]
@nqdtan fix in the new branch `2022.2.1`, if you get a chance, give it a try and let us know if there are any other issues.  Thanks!

**Reply by nqdtan** [reply]
This is working well without any issues. Thanks Chris!
